### PR TITLE
[CSM] Add callback on open inventory

### PR DESCRIPTION
**Pull request by Dumbeldor**
#5481

### CODE REVIEW

File: src/game.cpp [hunk 1]
```diff
@@ -2655,6 +2655,10 @@ void Game::openInventory()
 	InventoryLocation inventoryloc;
 	inventoryloc.setCurrentPlayer();
 	current_formspec->setFormSpec(fs_src->getForm(), inventoryloc);
+
+	if (client->moddingEnabled()) {
+		client->getScript()->on_open_inventory();
```
Comment (nerzhul) [inline src/game.cpp]: better usage, move this before  `current_formspec->setFormSpec` and return a bool, if false, open the standard formspec else don't open

File: builtin/client/register.lua [hunk 2]
```diff
@@ -71,3 +71,4 @@ core.registered_on_dignode, core.register_on_dignode = make_registration()
 core.registered_on_punchnode, core.register_on_punchnode = make_registration()
 core.registered_on_placenode, core.register_on_placenode = make_registration()
 core.registered_on_item_use, core.register_on_item_use = make_registration()
+core.registered_on_open_inventory, core.register_on_open_inventory = make_registration()
```
Comment (nerzhul) [inline builtin/client/register.lua]: on_open_inventory -> on_inventory_opening

File: src/game.cpp (hunk 1 above)
Comment (Dumbeldor) [inline src/game.cpp]: Yeah it can be useful nerzhul. It is done ! :)

File: src/game.cpp [hunk 3]
```diff
@@ -2647,14 +2647,21 @@ void Game::openInventory()
...
-	TextDest *txt_dst = new TextDestPlayerInventory(client);
-
-	create_formspec_menu(&current_formspec, client, device, &input->joystick, fs_src, txt_dst);
-	cur_formname = "";
 
 	InventoryLocation inventoryloc;
 	inventoryloc.setCurrentPlayer();
-	current_formspec->setFormSpec(fs_src->getForm(), inventoryloc);
+	bool showInventory = true;
+	if (client->moddingEnabled()) {
```
Comment (nerzhul) [inline src/game.cpp]: if (client->moddingEnabled() && client->getScript()->on_open_inventory(fs_src->m_client->getInventory(inventoryloc))) {

**Comment by Dumbeldor** [comment]
When no mod uses the callback the default returned value is false. As a result, the inventory is not displayed.
So I reverse. When the callback returns false I display the inventory.

File: src/game.cpp [hunk 4]
```diff
@@ -2647,14 +2647,15 @@ void Game::openInventory()
...
 	PlayerInventoryFormSource *fs_src = new PlayerInventoryFormSource(client);
-	TextDest *txt_dst = new TextDestPlayerInventory(client);
-
-	create_formspec_menu(&current_formspec, client, device, &input->joystick, fs_src, txt_dst);
-	cur_formname = "";
 
 	InventoryLocation inventoryloc;
 	inventoryloc.setCurrentPlayer();
-	current_formspec->setFormSpec(fs_src->getForm(), inventoryloc);
+	if (!client->moddingEnabled() || !client->getScript()->on_open_inventory(fs_src->m_client->getInventory(inventoryloc))) {
```
Comment (nerzhul) [inline src/game.cpp]: line too long

File: clientmods/preview/init.lua [hunk 5]
```diff
@@ -15,6 +15,12 @@ core.register_on_connect(function()
 	print("Server port: " .. server_info.port)
 end)
 
+core.register_on_inventory_opening(function(inventory)
+	print("INVENTORY OPEN")
+	print(dump(inventory))
+    return false
```
Comment (SmallJoker) [inline clientmods/preview/init.lua]: wrong indent.

File: builtin/client/register.lua [hunk 6]
```diff
@@ -71,3 +71,4 @@ core.registered_on_dignode, core.register_on_dignode = make_registration()
 core.registered_on_punchnode, core.register_on_punchnode = make_registration()
 core.registered_on_placenode, core.register_on_placenode = make_registration()
 core.registered_on_item_use, core.register_on_item_use = make_registration()
+core.registered_on_open_inventory, core.register_on_inventory_opening = make_registration()
```
Comment (SmallJoker) [inline builtin/client/register.lua]: Should be `, core.register_on_inventory_open = ` for consistency.

File: src/script/common/c_content.cpp [hunk 7]
```diff
@@ -1209,7 +1209,7 @@ void push_inventory_list(lua_State *L, Inventory *inv, const char *name)
 	}
 	std::vector<ItemStack> items;
 	for(u32 i=0; i<invlist->getSize(); i++)
-		items.push_back(invlist->getItem(i));
+		items.push_back((ItemStack) invlist->getItem(i));
```
Comment (SmallJoker) [inline src/script/common/c_content.cpp]: Why this typecast?

Comment (Dumbeldor) [inline src/script/common/c_content.cpp]: The function returns either an ItemStack const or an ItemStack. 
It was for my IDE to be happy. I remove, but we could cast in (const ItemStack) right?

**Comment by bigfoot547** [comment]
Can this be reviewed again?

File: src/game.cpp [hunk 8]
```diff
@@ -2675,14 +2675,17 @@ void Game::openInventory()
...
-
-	create_formspec_menu(&current_formspec, client, &input->joystick, fs_src, txt_dst);
-	cur_formname = "";
 
 	InventoryLocation inventoryloc;
 	inventoryloc.setCurrentPlayer();
-	current_formspec->setFormSpec(fs_src->getForm(), inventoryloc);
+
+	if (!client->moddingEnabled()
+			|| !client->getScript()->on_inventory_open(fs_src->m_client->getInventory(inventoryloc))) {
```
Comment (red-001) [inline src/game.cpp]: can't you access the client more directly using `client->`?

File: src/script/cpp_api/s_client.cpp [hunk 9]
```diff
@@ -224,6 +224,27 @@ bool ScriptApiClient::on_item_use(const ItemStack &item, const PointedThing &poi
...
+bool ScriptApiClient::on_inventory_open(Inventory *inventory)
+{
+	SCRIPTAPI_PRECHECKHEADER
+
+	lua_getglobal(L, "core");
+	lua_getfield(L, -1, "registered_on_inventory_open");
+
+	std::vector<const InventoryList*> lists = inventory->getLists();
+	std::vector<const InventoryList*>::iterator iter = lists.begin();
+	lua_createtable(L, 0, lists.size());
```
Comment (red-001) [inline src/script/cpp_api/s_client.cpp]: either add the code for pushing an inventory to lua to `c_content.cpp` and document this in `client_lua_api` or modify `l_inventory.cpp` to work on the client.

File: src/game.cpp (hunk 8 above)
Comment (Dumbeldor) [inline src/game.cpp]: Yeah you are right ! ;)

File: src/script/common/c_content.cpp [hunk 10]
```diff
@@ -1250,6 +1250,19 @@ void read_inventory_list(lua_State *L, int tableindex,
 	}
 }
 
+void push_inventory_lists(lua_State *L, Inventory *inventory)
+{
+	std::vector<const InventoryList*> lists = inventory->getLists();
+	std::vector<const InventoryList*>::iterator iter = lists.begin();
+	lua_createtable(L, 0, lists.size());
+	for (; iter != lists.end(); iter++) {
```
Comment (nerzhul) [inline src/script/common/c_content.cpp]: for range based loop please

File: src/script/common/c_content.h [hunk 11]
```diff
@@ -118,6 +118,9 @@ void               read_inventory_list       (lua_State *L, int tableindex,
                                               Inventory *inv, const char *name,
                                               Server *srv, int forcesize=-1);
 
+void               push_inventory_lists      (lua_State *L,
+											  Inventory *inventory);
```
Comment (nerzhul) [inline src/script/common/c_content.h]: crazy indent

File: doc/client_lua_api.md [hunk 12]
```diff
@@ -683,6 +683,10 @@ Call these functions only at load time!
     * Called when the local player uses an item.
     * Newest functions are called first.
     * If any function returns true, the item use is not sent to server.
+* `minetest.register_on_inventory_open(func(inventory))`
+    * Called when the local player open inventory
```
Comment (rubenwardy) [inline doc/client_lua_api.md]: Opens the Inventory

File: doc/client_lua_api.md [hunk 13]
```diff
@@ -683,6 +683,10 @@ Call these functions only at load time!
...
+    * Newest functions are called first
```
Comment (rubenwardy) [inline doc/client_lua_api.md]: This is inconsistent. Why?

File: src/game.cpp [hunk 14]
```diff
@@ -2667,14 +2667,17 @@ void Game::openInventory()
...
-
-	create_formspec_menu(&current_formspec, client, &input->joystick, fs_src, txt_dst);
-	cur_formname = "";
 
 	InventoryLocation inventoryloc;
 	inventoryloc.setCurrentPlayer();
-	current_formspec->setFormSpec(fs_src->getForm(), inventoryloc);
+
+	if (!client->moddingEnabled()
+			|| !client->getScript()->on_inventory_open(client->getInventory(inventoryloc))) {
```
Comment (rubenwardy) [inline src/game.cpp]: Break after binary op

File: doc/client_lua_api.md (hunk 13 above)
Comment (Dumbeldor) [inline doc/client_lua_api.md]: I don't understand sorry :(

**Pull request merged**